File: backend/src/main/java/com/dogratech/backend/service/TaskService.java
package com.dogratech.backend.service;

import com.dogratech.backend.dto.TaskRequestDTO;
import com.dogratech.backend.dto.TaskResponseDTO;
import com.dogratech.backend.enums.TaskStatus;
import org.springframework.data.domain.Page;

public interface TaskService {

    Page<TaskResponseDTO> getAllTasks(
            TaskStatus status,
            int page,
            int size,
            String sortBy,
            String direction
    );

    TaskResponseDTO getTaskById(Long id);

    TaskResponseDTO createTask(TaskRequestDTO dto);

    TaskResponseDTO updateTask(Long id, TaskRequestDTO dto);

    void deleteTask(Long id);
}

File: frontend/src/app/services/task.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Task, TaskRequest, PagedResponse, TaskParams } from '../models/task.model';

@Injectable({ providedIn: 'root' })
export class TaskService {
  private readonly API = 'http://localhost:8080/api/tasks';

  constructor(private http: HttpClient) {}

  getTasks(params: TaskParams = {}): Observable<PagedResponse<Task>> {
    let httpParams = new HttpParams();
    if (params.page !== undefined) httpParams = httpParams.set('page', params.page);
    if (params.size !== undefined) httpParams = httpParams.set('size', params.size);
    if (params.sortBy) httpParams = httpParams.set('sortBy', params.sortBy);
    if (params.direction) httpParams = httpParams.set('direction', params.direction);
    if (params.status) httpParams = httpParams.set('status', params.status);
    return this.http.get<PagedResponse<Task>>(this.API, { params: httpParams });
  }

  getTaskById(id: number): Observable<Task> {
    return this.http.get<Task>(`${this.API}/${id}`);
  }

  createTask(task: TaskRequest): Observable<Task> {
    return this.http.post<Task>(this.API, task);
  }

  updateTask(id: number, task: TaskRequest): Observable<Task> {
    return this.http.put<Task>(`${this.API}/${id}`, task);
  }

  deleteTask(id: number): Observable<string> {
    return this.http.delete(`${this.API}/${id}`, { responseType: 'text' });
  }
}
